### DISCUSSION TITLE
Subscribe to 700 data change

### DISCUSSION THREAD

**Discussion by bartols** [open62541pp/open62541pp#417] [Q&A] [answered]
Hi everybody I have to subscribe to among 700 nodes in an embedded system and I'm wondering which method is the more idiomatic and better. Or may be have to use the open62541 C lib.
Thanks in advance

**Comment by lukasberbuer** [comment]
Hi @bartols,
don't use the `Node` class but the raw service in the `services` namespace to create data change notifications for multiple nodes.

```cpp
[[nodiscard]] CreateMonitoredItemsResponse createMonitoredItemsDataChange(
    Client& connection,
    const CreateMonitoredItemsRequest& request,
    DataChangeNotificationCallback dataChangeCallback,
    DeleteMonitoredItemCallback deleteCallback = {}
);
```

https://open62541pp.github.io/open62541pp/group__CreateMonitoredItems.html#gad8948aef3e5e86d2b0a7996469e09b04 

Please let me know if you have further questions.

**Reply by bartols** [reply]
Thanks @lukasberbuer I tried but doesn't work may be I missed something:

    opcua::Client client;

    const opcua::MonitoredItemCreateRequest item1(
        {{1, "0:BIT_ALW_OFF"}, opcua::AttributeId::Value});
    const opcua::MonitoredItemCreateRequest item2({{1, "0:BIT_ALW_ON"}, opcua::AttributeId::Value});

    const opcua::CreateMonitoredItemsRequest request({}, 1U, opcua::TimestampsToReturn::Both,
                                                     {item1, item2});

    auto callback = [&](uint32_t subId, uint32_t monId, const opcua::DataValue &value) -> void {
        opcua::MonitoredItem item(client, subId, monId);
        std::cout << "Data change notification:\n"
                  << "- subscription id:   " << item.subscriptionId() << "\n"
                  << "- monitored item id: " << item.monitoredItemId() << "\n"
                  << "- node id:           " << item.getNodeId().toString() << "\n";
    };

    auto callbackDelete = [&](uint32_t subId, uint32_t monId) -> void {
        opcua::MonitoredItem item(client, subId, monId);
        std::cout << "Delete notification:\n"
                  << "- subscription id:   " << item.subscriptionId() << "\n"
                  << "- monitored item id: " << item.monitoredItemId() << "\n"
                  << "- node id:           " << item.getNodeId().toString() << "\n";
    };

    client.onSessionActivated([&] {
        std::cout << "SessionActivated\n";
        auto result = opcua::services::createMonitoredItemsDataChange(client, request, callback,
                                                                      callbackDelete);
    });

    while (true) {
        try {
            client.connect(OPC_SERVER);
            // Run the client's main loop to process callbacks and events.
            // This will block until client.stop() is called or an exception is thrown.
            client.run();
        } catch (const opcua::BadStatus &e) {
            if (e.code() == UA_STATUSCODE_BADNODEIDUNKNOWN) {
                // a not existing node
                std::cout << "Node not present on server: " << e.what() << "\n";
            }
            else {
                client.disconnect();
                std::cout << "Error: " << e.what() << "\nRetry to connect in 3 seconds\n";
                std::this_thread::sleep_for(std::chrono::seconds(3));
            }
        }
    }

**Reply by lukasberbuer** [reply]
Can you describe the error or what doesn't work?

**Reply by bartols** [reply]
No changes are signaled, the callback is never be called.
In 
`auto result = opcua::services::createMonitoredItemsDataChange(client, request, callback,
                                                                  callbackDelete);`
result is empty

**Reply by lukasberbuer** [reply]
Ok, there is a lot to cover:

First, you should check the status code of the response:
```cpp
    client.onSessionActivated([&] {
        std::cout << "SessionActivated\n";
        const auto response = opcua::services::createMonitoredItemsDataChange(
            client, request, callback, callbackDelete
        );
        response.getResponseHeader().getServiceResult().throwIfBad();
    });
```

This will tell you the error:
```
Error: BadSubscriptionIdInvalid
```

You can not create a MonitoredItem without a Subscription (https://reference.opcfoundation.org/Core/Part4/v105/docs/5.13). Let's fix that:

```cpp
    client.onSessionActivated([&] {
        std::cout << "SessionActivated\n";

        std::cout << "Create subscription\n";
        const opcua::SubscriptionParameters parameters{};
        const auto createSubscriptionResponse = opcua::services::createSubscription(
            client,
            parameters,
            true,  // publishingEnabled
            {},  // statusChangeCallback
            {}  // deleteCallback
        );
        createSubscriptionResponse.getResponseHeader().getServiceResult().throwIfBad();
        const auto subId = createSubscriptionResponse.getSubscriptionId();

        std::cout << "Create monitored item\n";
        const opcua::CreateMonitoredItemsRequest request(
            {},
            subId,  // <-- use subId from createSubscriptionResponse
            opcua::TimestampsToReturn::Both,
            {
                item1,
                item2,
            }
        );
        const auto createMonitoredItemResponse = opcua::services::createMonitoredItemsDataChange(
            client, request, callback, callbackDelete
        );
        createMonitoredItemResponse.getResponseHeader().getServiceResult().throwIfBad();
    });
```

The full example looks like this:
```cpp
#include <chrono>
#include <iostream>
#include <thread>

#include <open62541pp/client.hpp>
#include <open62541pp/services/monitoreditem.hpp>
#include <open62541pp/services/subscription.hpp>

int main() {
    opcua::Client client;

    const opcua::MonitoredItemCreateRequest item1({{1, "0:BIT_ALW_OFF"}, opcua::AttributeId::Value}
    );
    const opcua::MonitoredItemCreateRequest item2({{1, "0:BIT_ALW_ON"}, opcua::AttributeId::Value});

    auto callback =
        [&](uint32_t subId, uint32_t monId, [[maybe_unused]] const opcua::DataValue& value
        ) -> void {
        opcua::MonitoredItem item(client, subId, monId);
        std::cout
            << "Data change notification:\n"
            << "- subscription id:   " << item.subscriptionId() << "\n"
            << "- monitored item id: " << item.monitoredItemId() << "\n"
            << "- node id:           " << item.getNodeId().toString() << "\n";
    };

    auto callbackDelete = [&](uint32_t subId, uint32_t monId) -> void {
        opcua::MonitoredItem item(client, subId, monId);
        std::cout
            << "Delete notification:\n"
            << "- subscription id:   " << item.subscriptionId() << "\n"
            << "- monitored item id: " << item.monitoredItemId() << "\n"
            << "- node id:           " << item.getNodeId().toString() << "\n";
    };

    client.onSessionActivated([&] {
        std::cout << "SessionActivated\n";

        std::cout << "Create subscription\n";
        const opcua::SubscriptionParameters parameters{};
        const auto createSubscriptionResponse = opcua::services::createSubscription(
            client,
            parameters,
            true,  // publishingEnabled
            {},  // statusChangeCallback
            {}  // deleteCallback
        );
        createSubscriptionResponse.getResponseHeader().getServiceResult().throwIfBad();
        const auto subId = createSubscriptionResponse.getSubscriptionId();

        std::cout << "Create monitored item\n";
        const opcua::CreateMonitoredItemsRequest request(
            {},
            subId,  // <-- use subId from createSubscriptionResponse
            opcua::TimestampsToReturn::Both,
            {
                item1,
                item2,
            }
        );
        const auto createMonitoredItemResponse = opcua::services::createMonitoredItemsDataChange(
            client, request, callback, callbackDelete
        );
        createMonitoredItemResponse.getResponseHeader().getServiceResult().throwIfBad();
    });

    while (true) {
        try {
            client.connect("opc.tcp://localhost:4840");
            // Run the client's main loop to process callbacks and events.
            // This will block until client.stop() is called or an exception is thrown.
            client.run();
        } catch (const opcua::BadStatus& e) {
            if (e.code() == UA_STATUSCODE_BADNODEIDUNKNOWN) {
                // a not existing node
                std::cout << "Node not present on server: " << e.what() << "\n";
            } else {
                client.disconnect();
                std::cout << "Error: " << e.what() << "\nRetry to connect in 3 seconds\n";
                std::this_thread::sleep_for(std::chrono::seconds(3));
            }
        }
    }
}
```

**Reply by bartols** [reply]
Thank you @lukasberbuer now is perfect, a last thing, if in the list of monitored nodeIds there is one that is not correct (a not existing nodeId) the delete callback is called but monitoredItemId is 0 and not is possible to know which one is not correct.
The situation is: we have to monitor 700 different nodes and is easy that someone mispelled some of them we'd like to log the not existing node. Thanks in advance

**Reply by lukasberbuer** [reply]
You can check the result (`CreateMonitoredItemResult`) for each monitored item within the response:
https://open62541pp.github.io/open62541pp/classopcua_1_1CreateMonitoredItemsResponse.html
Each `CreateMonitoredItemResult` stores a `statusCode` to check for errors.

**Comment by bartols** [comment]
Thank you for your kind reply, sorry if I use you like an OP CUA expert but
I have to read variables with OP CUA from a system, transfer to another and
write them.
Basically a collector of homogeneous systems.
There is something yet available to serialize nodeIds and values to be
shared between applications?
Thanks in advance

**Reply by lukasberbuer** [reply]
I don't fully understand the use case. Let me try to summarize: You have two clients. One client is reading/subscribing variables nodes from an OPC UA server. The second client writes values to those variable nodes. Now you need to exchange the information between the clients? Just the node ids? If the second client can write those nodes, why not read them directly there?

**Comment by bartols** [comment]
Yes one client read nodeid and values send them to the other client that
will write in another device. Practically is how to serialize nodeId and
values, I tried this but doesn't work:

typedef struct {
    UA_NodeId nodeId;
    UA_Variant value;
} NodeData;

TEST(SerializerTest, Serialize)
{
    UA_ByteString byteString;
    UA_ByteString_init(&byteString);

    {
        double d = 3.0;

        NodeData data;
        data.nodeId = UA_NODEID_NUMERIC(3, 1003);
        UA_Variant_init(&data.value);
        UA_Variant_setScalar(&data.value, &d, &UA_TYPES[UA_TYPES_DOUBLE]);

        printf("%d,%d\n", data.nodeId.namespaceIndex,
data.nodeId.identifier.numeric);
        printf("Double: %f\n", *(double *)data.value.data);

        size_t totalSize = 0;
        totalSize += UA_calcSizeBinary(&data.nodeId,
&UA_TYPES[UA_TYPES_NODEID]);
        totalSize += UA_calcSizeBinary(&data.value,
&UA_TYPES[UA_TYPES_VARIANT]);

        byteString.data   = (UA_Byte *)UA_malloc(totalSize);
        byteString.length = (UA_Int32)totalSize;

        UA_encodeBinary(&data.nodeId, &UA_TYPES[UA_TYPES_NODEID],
&byteString);
        UA_encodeBinary(&data.value, &UA_TYPES[UA_TYPES_VARIANT],
&byteString);
    }

    {
        NodeData data_res;
        UA_decodeBinary(&byteString, &data_res.nodeId,
&UA_TYPES[UA_TYPES_NODEID], NULL);
        UA_decodeBinary(&byteString, &data_res.value,
&UA_TYPES[UA_TYPES_VARIANT], NULL);

        printf("%d,%d\n", data_res.nodeId.namespaceIndex,
data_res.nodeId.identifier.numeric);
        printf("Double: %f\n", *(double *)data_res.value.data);
    }
}

**Reply by lukasberbuer** [reply]
Maybe a stupid question: Why not do everything within one client?

**Comment by bartols** [comment, marked as answer]
Because they are on different machine. I find how to serialize:

```cpp
struct NodeData {
    opcua::NodeId nodeId;
    opcua::Variant value;
};

TEST(SerializerTest, Serialize)
{
    UA_ByteString byteString;
    UA_ByteString_init(&byteString);

    {
        double d = 3.0;

        NodeData data;
        data.nodeId = {3, 1003};
        data.value.setScalarCopy<double>(d);

        printf("%s -> %f\n", data.nodeId.toString().c_str(), data.value.getScalarCopy<double>());

        size_t totalSize = 0;
        totalSize += UA_calcSizeBinary(data.nodeId.handle(), &UA_TYPES[UA_TYPES_NODEID]);
        totalSize += UA_calcSizeBinary(data.value.handle(), &UA_TYPES[UA_TYPES_VARIANT]);

        byteString.data   = (UA_Byte *)UA_malloc(totalSize);
        byteString.length = (UA_Int32)totalSize;

        UA_Byte *bufPos       = byteString.data;
        const UA_Byte *bufEnd = &byteString.data[byteString.length];
        UA_StatusCode retval  = 0;
        retval |= UA_NodeId_encodeBinary(data.nodeId.handle(), &bufPos, bufEnd);
        retval |= UA_Variant_encodeBinary(data.value.handle(), &bufPos, bufEnd);
        ASSERT_EQ(retval, 0);
    }

    {
        NodeData data_res;
        UA_StatusCode retval = 0;
        size_t offset        = 0;
        retval |= UA_NodeId_decodeBinary(&byteString, &offset, data_res.nodeId.handle());
        retval |= UA_Variant_decodeBinary(&byteString, &offset, data_res.value.handle());

        ASSERT_EQ(retval, 0);

        ASSERT_EQ(data_res.nodeId->namespaceIndex, 3);
        ASSERT_EQ(data_res.nodeId->identifier.numeric, 1003);
        ASSERT_EQ(data_res.value.getScalarCopy<double>(), 3.0);

        printf("%s\n", data_res.nodeId.toString().c_str());
        printf("%f\n", data_res.value.getScalarCopy<double>());
    }
}

```

**Reply by lukasberbuer** [reply]
Ok understood 😊
Yes the way with `UA_encodeBinary`/`UA_decodeBinary` should work well.
Do you use the C library or open62541pp? We could add free functions `opcua::encodeBinary` and `opcua::decodeBinary` that simplifies the handling, especially with the `TypeRegistry` (no need to specify the `UA_DataType` anymore).

**Comment by bartols** [comment]
They would be useful, if I find a spare time could try to do and make a
pull request

Il Mer 6 Nov 2024, 18:01 Lukas Berbuer ***@***.***> ha
scritto:

**Reply by lukasberbuer** [reply]
That would be awesome 👍
